lib/terminus.js

```diff
@@ -1,143 +1,171 @@
+   * system signals and decorating the given server with
+   * the given health checks.
+   * @param {any} server The HTTP server instance which should get decorated with the given healthchecks
+   * @param {object} options The options which configure the Termninus instance
+   */
+  constructor (server, options = {}) {
+    this._server = server
+    this._options = Object.assign(DEFAULT_OPTIONS, options)
+    this._state = {
+      isShuttingDown: false
+    }
+
+    this._onSignal = options.onSignal || options.onSigterm || noopResolves
 
-  res.statusCode = 200
-  res.setHeader('Content-Type', 'application/json')
-  if (info) {
-    return res.end(JSON.stringify({
-      status: 'ok',
-      info: info
-    }))
+    if (Object.keys(this._options.healthChecks).length > 0) {
+      this._decorateWithHealthCheck()
+    }
+
+    // push the signal into the array
+    // for backwards compatability
+    if (!this._options.signals.includes(this._options.signal)) {
+      this._options.signals.push(this._options.signal)
+    }
+    this._decorateWithSignalHandler()
   }
-  res.end(SUCCESS_RESPONSE)
-}
 
-async function sendFailure (res, options) {
-  const { error, onSendFailureDuringShutdown } = options
+  async _cleanup (signal) {
+    if (!this._state.isShuttingDown) {
+      this._state.isShuttingDown = true
+      try {
+        await this._options.beforeShutdown()
+        await this._asyncServerStop()
+        await this._onSignal()
+        await this._options.onShutdown()
+        this._options.signals.forEach(sig => process.removeListener(sig, () => this._cleanup()))
+        process.kill(process.pid, signal)
+      } catch (error) {
+        this._options.logger('error happened during shutdown', error)
+        process.exit(1)
+      }
+    }
+  }
 
-  if (onSendFailureDuringShutdown) {
-    await onSendFailureDuringShutdown()
+  _sendSuccess (info) {
+    const statusResponse = { status: 'ok' }
+    if (info) {
+      statusResponse.info = info
+    }
+    return statusResponse
   }
-  res.statusCode = 503
-  res.setHeader('Content-Type', 'application/json')
-  if (error) {
-    return res.end(JSON.stringify({
+
+  _sendFailure (error) {
+    return {
       status: 'error',
       error: error
-    }))
+    }
   }
-  res.end(FAILURE_RESPONSE)
-}
 
-const intialState = {
-  isShuttingDown: false
-}
+  _executeHealtCheck (url) {
+    return this._options.healthChecks[url]()
+  }
 
-function noop () {}
+  async _onRequest (req, res, listener) {
+    let statusResponse
 
-function decorateWithHealthCheck (server, state, options) {
-  const { healthChecks, logger, onSendFailureDuringShutdown } = options
+    try {
+      statusResponse = await this.getHealthStatus(req.url)
+    } catch (error) {
+      listener(req, res)
+    }
 
-  server.listeners('request').forEach((listener) => {
-    server.removeListener('request', listener)
-    server.on('request', async (req, res) => {
-      if (healthChecks[req.url]) {
```

Comment:
this logic seems to be gone from the refactor; to check if there is a  healthcheck registered to the url or not

if we don't add it back, our logic will run for all the routes, and an error will be thrown (which will be try-catched, but it has performance implications)

Reply:
I think the behavior would not change after my PR, but you're right it is not as performant. Added back in.